proposals/NNNN-resource-instance-analysis.md

```diff
@@ -0,0 +1,71 @@
+* Proposal: [NNNN - Resource Instance Analysis](NNNN-resource-instance-analysis.md)
+* Author(s): [Ashley Coleman](https://github.com/V-FEXrt)
+* Status: **Design In Progress**
+
+## Introduction
+
+This proposal introduces a new term, Resource Instance, which represents a 
+unique and specific resource. The instance is referenced via a handle returned
+from the create handle intrinsics and it is possible to have multiple handles
+pointing at a specific Resource Instance, but each unique instance is
+disambiguated by the handle creation parameters.
+
+Beyond creation parameters, Reource Instances have several pieces of associated
+metadata. Type, CounterDirection (HasCounter), and GloballyCoherent are
+identified examples. Type is currently handled well so this proposal suggests
+no changes. However CounterDirection and GloballyCoherent require careful late 
+stage analysis to derive the correct metadata.
+
+To serve the needs of CounterDirection and to setup a framework for future
+instance metadata this document proposes that instance analysis be combined
+into the DXILResourceBindingAnalysis pass and that the pass be renamed.
+
+
+## Proposed solution
+
+In order to support the generic resource instance analysis `DXILResourceBindingAnalysis` 
+and `ResourceBindingInfo` will be renamed to `DXILResourceAnalysis` and `ResourceInfo`
+respectively. Then associated instance metadata will be resolved during the
+`DXILResourceBindingAnalysis` pass.
+
+
+The following mechanical changes are required to align existing and inflight code
+to the new structure.
+
+- `CounterDirection` will supercede `HasCounter` which will be removed from `TypeInfo`
+- `CounterDirection` and `GloballyCoherent` will be added to `ResourceInfo`
+- `DXILResourceAnalysis` will have a new step to resolve the instance's counter direction.
+
+Certain generated DXIL can create an illegal state for the instance metadata
+such as an instance with both an incremented and decremented counter. When this
+occurs the analysis pass should prioritize performance for the common case and
+defer detailed error message calculation for the uncommon failure code path.
+
+To achieve that goal, a step in `DXILResourceAnalysis` may set a terminal or
+invalid value in the ResourceInfo that a later pass `DXILPostOptimizationValidationPass`
```

Comment:
> DXILPostOptimizationValidationPass

Does this pass exist already, or is this a new one that'll be added?

Reply:
New pass that is needed. I'll add that in